src/fsharp/TypeChecker.fs

```diff
@@ -11450,10 +11450,10 @@ and TcIncrementalLetRecGeneralization cenv scopem
         // pathological situations
         let freeInUncheckedRecBinds = 
             lazy ((emptyFreeTyvars, cenv.recUses.Contents) ||> Map.fold (fun acc vStamp _ -> 
-                       if uncheckedRecBindsTable.ContainsKey vStamp then 
-                           let fwdBind = uncheckedRecBindsTable.[vStamp]  
+                       match Map.tryFind vStamp uncheckedRecBindsTable with
+                       | Some fwdBind ->
```

Comment:
We prevent a double lookup, but this allocates. Not saying that preventing a double lookup is wrong, I'm just noting that this will allocate.

Reply:
I would feel better if we just use ImmutableDictionary instead of Map, and call TryGetValue, as it uses an out byref and we won't allocate when getting something from a dictionary.

Reply:
@TIHan One reason not to take a deep dependency on System.Collections.Immutable because that will make cross-compiling to Fable/JS harder

But you are right, we should really add a TryGetValue to the Map type we use.  There is a separate Zmap/TaggedCollections type in the compiler and we could make that our internal mutable collection type.

Reply:
Another reason not to use ImmutableDictionary is that AFAICS (last time i looked) it has poor node sharing characteristics on incremental update.  That is critical for some (but not all) of the internal dictionaries.

Reply:
Ok, that sounds good to me. Then looking at the Zmap/TaggedCollections might be a better choice.

Aside from that, if ImmutableDictionary has problems, we should report it to make it better.